### ISSUE TITLE
Support tofu plan -dry-run with mocked providers

### ISSUE THREAD

**Issue by gw0** [opentofu/opentofu#2159]
### OpenTofu Version

```shell
All
```

### The problem in your OpenTofu project

In order to combat configuration drift and gain insight into what consequences a change might have we attempted to implement GitOps principles in a safe manner. The Git repository contains the desired state of the infrastructure and if all changes are performed through OpenTofu/Terraform, then the desired state in Git should match the live state. In a mono-repo approach where Terragrunt and TF modules are heavily used, it is less obvious what a change might affect.

The plan was to have a CI pipeline that does not need access to the infrastructure or TF state, but can still show the consequences a change in a PR will have by running `tofu plan` in dry-run mode with empty TF state on the PR branch and compare its output with on the main branch. With other words, for every PR we want a diff between the old and new desired state or dry-run plan (not a diff against live state or TF state). Because we are just comparing the desired states, the CI pipeline does not need access to the infrastructure not even TF state. Of course, it does not fully test the entire provisioning, but it should be extremely fast and could run in parallel.

### Attempted Solutions

On a toy example we managed to produce "dry-run plan" with `tofu plan -refresh=false` and overriding the TF state with an empty file. The toy example even worked with Terragrunt with:

```sh
$ terragrunt plan -refresh=false --terragrunt-fetch-dependency-output-from-state=true
```

Unfortunately, most TF providers (AWS, GCP) want to initialize and require valid credentials. Even with workarounds the problem resurfaces in TF modules and submodules... On the other hand, OpenTofu 1.8 implemented mock providers for `tofu test` ([docs](https://opentofu.org/docs/intro/whats-new/#provider-mocking-in-tofu-test)) that "allows you to automatically generate mock values for resources and data sources on a per-provider basis" which is exactly what one would need for this use case.

### Proposal

Please add a `tofu plan -dry-run=true` (or similar) that would use `mock_provider` (no need for credentials) and produce a dry-run plan as if all resources would be generated from scratch (as with an empty TF state).

**Comment by ollevche** [comment]
Hello and thank you for this issue! The core team regularly reviews new issues and discusses them, but this can take a little time. Please bear with us while we get to your issue. If you're interested, the contribution guide has a section about the [decision-making process](https://github.com/opentofu/opentofu/blob/main/CONTRIBUTING.md#who-decides-if-a-feature-will-be-implemented-and-how-is-that-decision-made).

**Comment by gw0** [comment]
> OpenTofu tends to generate a highly optimistic diff that doesn't properly reflect what would happen if the changes were really applied

Exactly, I am asking for a highly optimistic "dry plan" to indicate what effects a change in a submodule might have, it will not be far from an exact `plan` and it does not need to provide any diff functionality. Just a way to print out the current configuration of all resources with all submodules, preferably in the `plan` format as if one would create all the resources from scratch. Or alternatively, a way to render the current configuration of all resources with all submodules after resolving as many of the input variables, conditional expressions, and removal of skipped resources (`count=0`). An equivalent from Kubernetes would be to render all final K8s manifests produced by Helm charts and Kustomize.
